lib/Transforms/Secretize/WrapGeneric.cpp

```diff
@@ -0,0 +1,92 @@
+#include <iostream>
+
+#include "include/Dialect/Secret/IR/SecretDialect.h"
+#include "include/Dialect/Secret/IR/SecretOps.h"
+#include "include/Dialect/Secret/IR/SecretTypes.h"
+#include "include/Transforms/Secretize/Passes.h"
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/IRMapping.h"             // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_WRAPGENERIC
+#include "include/Transforms/Secretize/Passes.h.inc"
+
+struct WrapWithGeneric : public OpRewritePattern<func::FuncOp> {
+  WrapWithGeneric(mlir::MLIRContext *context)
+      : mlir::OpRewritePattern<func::FuncOp>(context) {}
+
+  LogicalResult matchAndRewrite(func::FuncOp op,
+                                PatternRewriter &rewriter) const override {
+    bool hasSecrets = false;
+
+    SmallVector<Type, 4> newInputs;
+    for (unsigned i = 0; i < op.getNumArguments(); i++) {
+      if (op.getArgAttr(i, secret::SecretDialect::kArgSecretAttrName) !=
+          nullptr) {
+        hasSecrets = true;
+        op.removeArgAttr(i, secret::SecretDialect::kArgSecretAttrName);
+
+        auto newTy = secret::SecretType::get(op.getArgument(i).getType());
+        op.getArgument(i).setType(newTy);  // Updates the block argument type.
+        newInputs.push_back(newTy);
+      }
+    }
+
+    if (!hasSecrets) {
+      // Match failure, no secret inputs.
+      return failure();
+    }
+
+    auto newOutputs = llvm::to_vector<6>(llvm::map_range(
+        op.getResultTypes(),
+        [](Type t) -> Type { return secret::SecretType::get(t); }));
+
+    op.setFunctionType(
+        FunctionType::get(getContext(), {newInputs}, {newOutputs}));
```

Comment:
Wouldn't this drop any function arguments that don't have the secret annotation? There's no "else" block to the if statement on line 28. This works for our case where we assume all inputs are secret, but in general I can imagine cases where we want to have a function with only some inputs secret.

I think it would be best to keep those arguments by appending to `newInputs` in the else case. The changes in https://github.com/google/heir/pull/308 mean you don't have to pass the non-secret args to the generic anymore, but you can still do it (meaning no other changes are needed in this file), and then after https://github.com/j2kun/heir/pull/7 the canonicalization pass will automatically remove them (see `RemoveNonSecretGenericArgs`).

Reply:
argh! yes, i originally had that in, until I had at some point switched APIs to the line above ` op.getArgument(i).setType(newTy)` where I wouldn't need to consider the non-secrets. 

it turns out, there isn't a good way of setting the output type without `setFunctionType`, and so yes, I need to use the correct `newInputs` here.

Reply:
fixed, added a testcase!